File: day_01/day01.jl
#!/usr/env julia
# Advent of Code 2019
# Day 1

data = parse.(Int,readlines("input"))
fuel_count(mass) = mass ÷ 3 - 2
recursive_fuel(mass) = mass > 0 ? mass + recursive_fuel(fuel_count(mass)) : 0

println("Module fuel requirement: $(sum(fuel_count.(data)))")
println("Full fuel requirement: $(sum(recursive_fuel.(fuel_count.(data))))")

File: day_01/day01.pl
#!/usr/bin/perl
# Advent of Code 2019
# Day 1

use strict;
use warnings;
use List::Util qw(sum);

open(my $in, "<", "input") or die "Can't open: $!";
chomp(my @mass = <$in>);
close $in;

sub fuel_count {
    my $num = shift;
    use integer;
    return $num / 3 - 2;
}

sub calculate_fuel_recursion {
    my $atomic_mass = shift;
    return $atomic_mass > 0 ? $atomic_mass + calculate_fuel_recursion(fuel_count($atomic_mass)) : 0;
}

print "Module fuel requirement: ", sum(map {fuel_count($_)} @mass), "\n";
print "Full fuel requirement: ", sum(map {calculate_fuel_recursion(fuel_count($_))} @mass), "\n";

